### PR TITLE
feat(acl) allow acl plugin to use a consumer without credentials

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Sometimes Kong is not the provider of credentials, but a third-party identity provider does that. E.g. with OpenID Connect and LDAP the credentials are stored in 3rd party IdP or directory.

ACL plugin on the other hand did only work with `ngx.ctx.authenticated_credential`, this patch adds a support to `ngx.ctx.authenticated_consumer` as well.

### CODE REVIEW

File: kong/plugins/acl/handler.lua [hunk 1]
```diff
@@ -30,9 +30,15 @@ function ACLHandler:access(conf)
...
-  if ngx.ctx.authenticated_credential then
+  if ngx.ctx.authenticated_consumer then
+    consumer_id = ngx.ctx.authenticated_consumer.id
+  end
+
+  if not consumer_id and ngx.ctx.authenticated_credential then
     consumer_id = ngx.ctx.authenticated_credential.consumer_id
-  else
+  end
+
```
Comment (kikito) [inline kong/plugins/acl/handler.lua]: Wouldn't it be simpler to write an `elseif` here?

```
if ngx.ctx.authenticated_consumer then
  consumer_id = ngx.ctx.authenticated_consumer.id
elseif ngx.ctx.authenticated_credential then
  consumer_id = ngx.ctx.authenticated_credential.consumer_id
end
```

Or even

```
local consumer_id = (ngx.ctx.authenticated_consumer and ngx.ctx.authenticated_consumer.id) or
                    (ngx.ctx.authenticated_credential and ngx.ctx.authenticated_credential.consumer_id)
```

(That second one might be a bit too much)

Comment (p0pr0ck5) [inline kong/plugins/acl/handler.lua]: I'd be in favor of the latter approach ^ though the parens aren't necessary I don't think.

Comment (bungle) [inline kong/plugins/acl/handler.lua]: Actually no. There is always change that `ngx.ctx.authenticated_credential` is set but ` `ngx.ctx.authenticated_credential.customer_id` results `nil`. For example right now I'm working this around by setting `ngx.ctx.authenticated_credential.customer_id`, but that feels wrong as I cannot set `ngx.ctx.authenticated_credential.id` as there is no credential. My approach works even if people do stupid things, and it doesn't hurt anything (maybe just one extra `nil` check).

Comment (bungle) [inline kong/plugins/acl/handler.lua]: What I mean here is that even if there is `ngx.ctx.authenticated_consumer` table or `ngx.ctx. authenticated_credential` table,  there is no gurantee that it has any properties. We could go on, and even argue that the current and this patched code will crash if `ngx.ctx.authenticated_consumer` or `ngx.ctx. authenticated_credential` are not `tables`.

**Comment by bungle** [comment]
I pushed a new version that also doesn't crash if the used context variables are not `table`s.

File: kong/plugins/acl/handler.lua [hunk 2]
```diff
@@ -1,47 +1,76 @@
...
+local set_header   = ngx.req.set_header
+local ipairs       = ipairs
+local type         = type
+local next         = next
+local empty        = {}
+
 
 local ACLHandler = BasePlugin:extend()
 
+
```
Comment (thibaultcha) [inline kong/plugins/acl/handler.lua]: It seems like this PR has many style changes. As we've agreed upon by the past, could we limit the scope of the changes in the PRs? Or if we want those changes in the same PR, could they be in separate commits at least? 

- `feat(acl) work with consumer`
- `style(acl) update to new style`
- `perf(acl) minor improvements`

File: kong/plugins/acl/handler.lua [hunk 3]
```diff
@@ -1,47 +1,76 @@
...
+  if not consumer_id then
+    local authenticated_credential = ctx.authenticated_credential
+    if authenticated_credential and type(authenticated_credential) == "table" then
+      consumer_id = authenticated_credential.consumer_id
+    end
+  end
+
+  if not consumer_id then
+    return responses.send_HTTP_FORBIDDEN(
+      "Cannot identify the consumer, add an authentication plugin to use the ACL plugin"
```
Comment (thibaultcha) [inline kong/plugins/acl/handler.lua]: aside note: welp, I never saw this before, but I don't think user-facing errors/responses should talk about the notion of "plugins". This should probably be logged for the Kong admin to see (`ngx.log(ngx.WARN, ...)` maybe), and the 403 response shown to the user should be blank. This is, of course, out of the scope of this PR and is definitely **not** a blocker for this PR imo :)

File: kong/plugins/acl/handler.lua [hunk 4]
```diff
@@ -1,47 +1,76 @@
...
   end
 
   -- Retrieve ACL
-  local acls, err = cache.get_or_set(cache.acls_key(consumer_id), nil,
-                                load_acls_into_memory, consumer_id)
+  local acls, err = cache.get_or_set(
+    cache.acls_key(consumer_id),
+    nil,
+    load_acls_into_memory,
+    consumer_id)
```
Comment (thibaultcha) [inline kong/plugins/acl/handler.lua]: Any reason for this change and the new line below? I think it was fine as it was (per the code style), except for the second line not aligning with the first argument. Such style changes should be part of another commit/PR anyways.

File: kong/plugins/acl/handler.lua [hunk 5]
```diff
@@ -60,14 +89,17 @@ function ACLHandler:access(conf)
   if next(conf.whitelist or empty) then
     if not next(acls or empty) then
       block = true
+
```
Comment (thibaultcha) [inline kong/plugins/acl/handler.lua]: ditto: all those style changes - while appreciated, it'd be great if they could at least be included in a separate commit :) Thanks!

File: kong/plugins/acl/handler.lua [hunk 6]
```diff
@@ -83,7 +115,9 @@ function ACLHandler:access(conf)
   for _, v in ipairs(acls) do
     table_insert(str_acls, v.group)
   end
-  ngx.req.set_header(constants.HEADERS.CONSUMER_GROUPS, table_concat(str_acls, ", "))
+
+  set_header(constants.HEADERS.CONSUMER_GROUPS, table_concat(str_acls, ", "))
```
Comment (thibaultcha) [inline kong/plugins/acl/handler.lua]: Such a change should also be separated in its own `perf(acl)` commit as well
